Documentation

HassanAbouelela opened the pull request:
This PR introduces a few changes:
- Add documentation, using sphinx and google style
- Add versioning and changelog system (more on this below)
- Action to deploy the docs to GH pages

You can see what the docs branch will look like [here](https://github.com/HassanAbouelela/bot-core/tree/docs)
and what the actual docs will look like [here](https://hassanabouelela.github.io/bot-core/).

## Versioning
So I've gone ahead and started working out the exact details of the versioning system. The plan is:

We'll follow semvar. Upon version bumping, we'll update the version in the poetry config, and create a new branch with that version. I may or may not have picked this system because it integrates perfectly with the changelog program.

From there, we can make releases with the specified version, but I've held off on that till we figure out the exact method we'll be releasing the package.

(P.S I've played around with the linkcode system for a while to try and get it working well. It seems to be working alright from my testing (you can see what I've tested in the linked docs), but I wouldn't be entirely shocked if it broke in the future. If you manage to break, please let me know)

Bluenix2 commented on botcore/__init__.py:
```diff
@@ -0,0 +1,9 @@
+from botcore import (
+    regex,
+)
+
+__all__ = [
+    regex,
+]
+
+__all__ = list(map(lambda module: module.__name__, __all__))
```
What's the point of this?

HassanAbouelela replied:
We expose certain modules, so that people using the package know what they can import, and more relevantly here for auto docs.

The reason that line is there is to convert to strings, the expected format. The reason I don't just hardcode the strings is is because they can be broken by refractings. Imports like this will throw an error instead of silently failing.

jchristgit replied:
Won't unknown modules in `__all__` raise a warning in linters though? I'm pretty sure I've seen linter warnings for that before.

HassanAbouelela replied:
If they do, it's not set up in this project. Regardless, I think actual imports work better for automatic refactorings and guaranteed failures, what do you think?

jchristgit replied:
I'm still 50/50 on this, I'm not sure which linter complained about this though :(

Bluenix2 commented on docs/README.md:
```diff
@@ -0,0 +1,44 @@
+
+The output will be in the [`/docs/build`](.) directory.
+
+Additionally, there are two helper tasks: `apidoc` and `builddoc`.
+`apidoc` is responsible for calling autodoc and generating docs from docstrings.
+`builddoc` generates the HTML site, and should be called after apidoc.
```
```suggestion
Additionally, there are two helper tasks: `apidoc` and `builddoc`:

- `apidoc` is responsible for calling autodoc and generating docs from docstrings.
- `builddoc` generates the HTML site, and should be called after apidoc.
```

HassanAbouelela replied:
This seems very nit picky, and I actually prefer it as one paragraph.

Bluenix2 commented on docs/README.md:
```diff
@@ -0,0 +1,44 @@
+
+
+## Docstrings
+To have your code probably documented, you need to do a couple of things:
```
Don't you mean properly?

HassanAbouelela replied:
Ah yes, that is a typo. Thanks

HassanAbouelela replied:
e465278e

Bluenix2 commented on docs/README.md:
```diff
@@ -0,0 +1,44 @@
+
+
+## Docstrings
+To have your code probably documented, you need to do a couple of things:
```
Oh wait is this for it to appear in Sphinx?

Maybe mention that like "For your code to show up in the automatically generated documentation..".

HassanAbouelela replied:
I'm really not sure why this clarification is necessary. I'm going to reword it I guess since I'm going to be editing this line, but I mean, this is a general guide on writing your docstrings...

HassanAbouelela replied:
e465278e

Bluenix2 commented on docs/README.md:
```diff
@@ -0,0 +1,44 @@
+2. [Write your docstring, using the Google docstring format][google]
+
+Refer to the [sphinx documentation][docstring-sections] for more information.
+
+
+[google]: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
+[docstring-sections]: https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections
+
+
```
You don't really have double empty lunes anywhere in Markdown

```suggestion

[google]: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
[docstring-sections]: https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections

```

HassanAbouelela replied:
What do you mean? These are invisible, I've done it this way to add a visual separator between content and footnotes for editors.

Bluenix2 commented on botcore/regex.py:
```diff
@@ -0,0 +1,21 @@
+"""
+Regex for discord server invites.
+
+:meta hide-value:
```
Is this for PyCharm?

HassanAbouelela replied:
No it's for Sphinx. If we didn't add this, the docs would have the entire regex written to it. You can see a list of all flags in the Sphinx documentation.

HassanAbouelela commented:
I realize I forgot to add the favicon, so I've now added that. I also updated the color of the doc entries after some bike shedding in dev core. Both changes are live in the linked docs.

jchristgit commented on .github/workflows/docs.yaml:
```diff
@@ -0,0 +1,51 @@
+      - uses: actions/checkout@v2
+
+      - name: Install Python Dependencies
+        uses: HassanAbouelela/actions/setup-python@setup-python_v1.0.0
```
Isn't there an official setup-python we can use here?

HassanAbouelela replied:
The setup-python action just sets up python, while the one from my actions repo does a bit more under the hood to set up poetry and caching.

There is a bit more context for why this is on my private account in dev-core, but this is essentially something we're planning on moving to everywhere eventually, pending a few things.

jchristgit commented on pyproject.toml:
```diff
@@ -15,8 +15,14 @@ packages = [
-python = "^3.9"
+python = "3.9.*"
 "discord.py" = "^1.7.2"
+Sphinx = "^4.3.1"
+tomli = "^1.2.2"
+GitPython = "^3.1.24"
+sphinx-autodoc-typehints = "^1.12.0"
+releases = "^1.6.3"
+furo = "^2021.11.23"
```
Shouldn't these be `dev-dependencies`? I'm not sure how `poetry` ships these upstream, but if we package, we probably don't need to have to install all these dependencies to use the repository?

HassanAbouelela replied:
We could do it either way to be honest. I'll just move them I guess.

Poetry has an upcoming unreleased feature which allows grouping dependency in custom groups (think dev, docs, prod, testing, etc), but unfortunately that's not out yet. Waiting for the day it makes it out of alpha (though we could use alpha if we wanted to 👀).

HassanAbouelela replied:
2196b4c

jchristgit reviewed (commented):
Looks good overall, thank you! Two small comments.

Bluenix2 reviewed (approved):
Thanks, this is a great start on the documentation.
